File: lib-esphome-client/src/test/java/io/github/arhor/esphome/client/async/noise/NoiseKeyMaterialTest.java
package io.github.arhor.esphome.client.async.noise;

import io.github.arhor.esphome.client.async.exception.EspHomeProtocolException;
import org.junit.jupiter.api.Test;

import java.util.Base64;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class NoiseKeyMaterialTest {

    @Test
    void decodesValidBase64Key() {
        final var raw = new byte[32];
        for (var i = 0; i < 32; i++) {
            raw[i] = (byte) (i + 1);
        }
        final var encoded = Base64.getEncoder().encodeToString(raw);

        assertArrayEquals(raw, NoiseKeyMaterial.decodeBase64(encoded));
    }

    @Test
    void rejectsBlankKey() {
        assertThrows(EspHomeProtocolException.class, () -> NoiseKeyMaterial.decodeBase64("  "));
    }

    @Test
    void rejectsWrongLength() {
        final var encoded = Base64.getEncoder().encodeToString(new byte[16]);
        final var ex = assertThrows(EspHomeProtocolException.class, () -> NoiseKeyMaterial.decodeBase64(encoded));
        assertTrue(ex.getMessage().contains("32 bytes"));
    }
}

File: lib-esphome-client/src/test/kotlin/io/github/arhor/esphome/client/internal/noise/NoiseKeyMaterialTest.kt
package io.github.arhor.esphome.client.internal.noise

import io.github.arhor.esphome.client.exception.EspHomeProtocolException
import java.util.Base64
import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertFailsWith

class NoiseKeyMaterialTest {

    @Test
    fun `decodes base64 32 byte key`() {
        val expected = ByteArray(32) { index -> index.toByte() }
        val encoded = Base64.getEncoder().encodeToString(expected)

        val decoded = NoiseKeyMaterial.decodeBase64(encoded)

        assertContentEquals(expected, decoded)
    }

    @Test
    fun `rejects missing and invalid key material`() {
        assertFailsWith<EspHomeProtocolException> {
            NoiseKeyMaterial.decodeBase64(null)
        }
        assertFailsWith<EspHomeProtocolException> {
            NoiseKeyMaterial.decodeBase64("not base64")
        }
        assertFailsWith<EspHomeProtocolException> {
            NoiseKeyMaterial.decodeBase64(Base64.getEncoder().encodeToString(ByteArray(31)))
        }
    }
}
